unison-runtime/src/Unison/Runtime/Stack.hs

```diff
@@ -399,374 +462,370 @@ dumpFP fp sz (F n _) = fp + sz - n
+    ustk :: {-# UNPACK #-} !(MutableByteArray (PrimState IO)),
+    bstk :: {-# UNPACK #-} !(MutableArray (PrimState IO) Closure)
+  }
+
+instance Show Stack where
+  show (Stack ap fp sp _ _) =
+    "Stack " ++ show ap ++ " " ++ show fp ++ " " ++ show sp
+
+type UElem = Int
+
+type USeg = ByteArray
+
+type BElem = Closure
+
+type BSeg = Array Closure
+
+type Elem = (UElem, BElem)
+
+type Seg = (USeg, BSeg)
+
+alloc :: IO Stack
+alloc = do
+  ustk <- newByteArray 4096
+  bstk <- newArray 512 BlackHole
+  pure $ Stack {ap = -1, fp = -1, sp = -1, ustk, bstk}
+{-# INLINE alloc #-}
+
+peek :: Stack -> IO Elem
+peek stk = do
+  u <- upeek stk
+  b <- bpeek stk
+  pure (u, b)
+{-# INLINE peek #-}
+
+bpeek :: Stack -> IO BElem
+bpeek (Stack _ _ sp _ bstk) = readArray bstk sp
+{-# INLINE bpeek #-}
+
+upeek :: Stack -> IO UElem
+upeek (Stack _ _ sp ustk _) = readByteArray ustk sp
+{-# INLINE upeek #-}
+
+peekOff :: Stack -> Off -> IO Elem
+peekOff stk i = do
+  u <- upeekOff stk i
+  b <- bpeekOff stk i
+  pure (u, b)
+{-# INLINE peekOff #-}
+
+bpeekOff :: Stack -> Off -> IO BElem
+bpeekOff (Stack _ _ sp _ bstk) i = readArray bstk (sp - i)
+{-# INLINE bpeekOff #-}
+
+upeekOff :: Stack -> Off -> IO UElem
+upeekOff (Stack _ _ sp ustk _) i = readByteArray ustk (sp - i)
+{-# INLINE upeekOff #-}
+
+-- | Store an unboxed value and null out the boxed stack at that location, both so we know there's no value there,
+-- and so garbage collection can clean up any value that was referenced there.
+upoke :: Stack -> UElem -> IO ()
+upoke stk@(Stack _ _ sp ustk _) u = do
+  bpoke stk BlackHole
+  writeByteArray ustk sp u
+{-# INLINE upoke #-}
+
+-- | Store a boxed value.
+-- We don't bother nulling out the unboxed stack,
+-- it's extra work and there's nothing to garbage collect.
+bpoke :: Stack -> BElem -> IO ()
+bpoke (Stack _ _ sp _ bstk) b = writeArray bstk sp b
+{-# INLINE bpoke #-}
+
+upokeOff :: Stack -> Off -> UElem -> IO ()
+upokeOff stk i u = do
+  bpokeOff stk i BlackHole
+  writeByteArray (ustk stk) (sp stk - i) u
+{-# INLINE upokeOff #-}
+
+bpokeOff :: Stack -> Off -> BElem -> IO ()
+bpokeOff (Stack _ _ sp _ bstk) i b = writeArray bstk (sp - i) b
+{-# INLINE bpokeOff #-}
+
+-- | Eats up arguments
+grab :: Stack -> SZ -> IO (Seg, Stack)
+grab (Stack _ fp sp ustk bstk) sze = do
+  uSeg <- ugrab
+  bSeg <- bgrab
+  pure $ ((uSeg, bSeg), Stack (fp - sze) (fp - sze) (sp - sze) ustk bstk)
+  where
+    ugrab = do
+      mut <- newByteArray bsz
+      copyMutableByteArray mut 0 ustk (bfp - bsz) bsz
+      seg <- unsafeFreezeByteArray mut
+      moveByteArray ustk (bfp - bsz) ustk bfp fsz
+      pure seg
+      where
+        bsz = bytes sze
+        bfp = bytes $ fp + 1
+        fsz = bytes $ sp - fp
+    bgrab = do
+      seg <- unsafeFreezeArray =<< cloneMutableArray bstk (fp + 1 - sze) sze
+      copyMutableArray bstk (fp + 1 - sze) bstk (fp + 1) fsz
+      pure seg
+      where
+        fsz = sp - fp
+{-# INLINE grab #-}
+
+ensure :: Stack -> SZ -> IO Stack
```

Comment:
I don't know how much this matters, but it should only be necessary to do a single check about whether to allocate more stack. The size of the `ustk` in bytes is 8x the size of `bstk`. So you can just do the `bstk` check and reallocate both if necessary. It doesn't make sense for them to ever be out of sync such that you'd reallocate one and not the other.

Reply:
Done :)